cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -200,6 +201,16 @@ class GcsOutputStream : public arrow::io::OutputStream {
     return internal::ToArrowStatus(stream_.last_status());
   }
 
+  Future<> CloseAsync() override {
+    if (closed_) return Status::OK();
+
+    auto self = std::dynamic_pointer_cast<GcsOutputStream>(shared_from_this());
+    auto deferred = [self = std::move(self)]() -> Status { return self->Close(); };
```

Comment:
This class is not thread-safe, so the fact that the current thread is handing a shared_ptr to another thread can only be safe if the caller of this function forgets about all the shared pointers to this class after this call.

I wonder how many potential bugs we could find by adding the following WARNING:

```cpp
 Future<> CloseAsync() override {
   auto self = shared_from_this();
   if (self.use_count() > 2) {
     ARROW_LOG(WARNING) << "Too many strong references to GcsOutputStream before CloseAsync() call";
   }
   if (closed_) return Status::OK();
   ...
```

In other words, the caller of `CloseAsync` should have unique ownership of `CloseAsync` so it can give the IO thread unique ownership for the `Close` call.

Without this, every use of the class after `CloseAsync` call is a data race. Thankfully, the shared_ptr destructor checks an atomic ref count, so there is not risk of double destructor calls.

Reply:
In even more words: safe use of this class relies on it being used by only one thread at once.

Reply:
> This class is not thread-safe

IMO, the `CloseAsync()` itself cannot be called concurrently or called multiple times. And `Close` will not called by dtor if the Defered function is still alive. If we concern about the `CloseAsync`  can add a `Future<>` and checks it during `Close` and `CloseAsync`

Reply:
I've check the s3fs, maybe it also doesn't prevent from `FinishPartUploadAfterFlush()` being called multiple times. It's possible to make gcs safe by holding a future here.

For gcs, maybe the code below is thread safe

```
  Status Close() override {
    if (closed_) {
      return Status::OK();
    }
    return CloseAsync().status();
  }

  Future<> CloseAsync() override {
    if (closed_) return Status::OK();
    if (this->close_future_.is_valid()) {
      return close_future_;
    }
    auto self = std::dynamic_pointer_cast<GcsOutputStream>(shared_from_this());
    auto deferred = [self = std::move(self)]() -> Status {
      self->stream_.Close();
      self->closed_ = true;
      return internal::ToArrowStatus(self->stream_.last_status());
    };
    ARROW_ASSIGN_OR_RAISE(close_future_,
                          io::internal::SubmitIO(io_context_, std::move(deferred)));
    return close_future_;
  }
```

Maybe this

Reply:
> IMO, the CloseAsync() itself cannot be called concurrently or called multiple times. 

It should not but it may be mistakenly called and the `if (closed_)` tests gives the impression that that code checks against that. We're relying on the users' discretion.

```cpp
  ASSERT_OK_AND_ASSIGN(auto output, fs->OpenOutputStream(path, /*metadata=*/{}));
  ASSERT_OK(output->Write(text.data(), text.size()));
  auto close_future = output->CloseAsync();
  // It's UB to use `output->...` from here.
  // `output->closed_` might be read as `false` by this thread
  // even after the I/O thread set it to `true`.
```

Reply:
I don't think you should complicate the code before concrete bug reports, hence my suggestion for the WARNING as a way to catch potential misuse.

Reply:
🤔 A called flag or a `Future<>` can be ok here. If `CloseAsync/Close` not being called after current `CloseAsync` it could be safe. Checking `shared_from_this()` is a bit hacking to be, think about something like Compressed Output Stream...

Reply:
Oh, another problem here is `bool closed()` is weird after this patch